File: frontend/src/financial/lib/qb-signals.ts
/**
 * Turns the QuickBooks overview payload into a short list of things a human
 * should actually do something about.
 *
 * The old dashboard rendered every number it had and left the reader to work
 * out which ones were bad. Everything here answers that question instead: a
 * signal only exists once a threshold is crossed, and it says so in a sentence.
 *
 * Pure functions, no React. Self-check at the bottom: `npx tsx src/financial/lib/qb-signals.ts`
 */

import type { QuickBooksOverview } from "../types/quickbooks";

export type Severity = "critical" | "warn" | "info";

export interface Signal {
  id: string;
  severity: Severity;
  /** Plain-English statement of the problem. No jargon, no metric names. */
  headline: string;
  /** The one number that sizes the problem. Pre-formatted. */
  figure?: string;
  /** Why it matters or what to do. One clause. */
  detail?: string;
  /** Tab id to jump to for the underlying rows. */
  goTo?: string;
}

const SEVERITY_RANK: Record<Severity, number> = { critical: 0, warn: 1, info: 2 };

const usd = (n: number) =>
  n.toLocaleString("en-US", {
    style: "currency",
    currency: "USD",
    maximumFractionDigits: 0,
  });

const plural = (n: number, one: string, many = `${one}s`) => (n === 1 ? one : many);

/** Buckets QuickBooks reports as genuinely late, not merely outstanding. */
const LATE_BUCKETS = new Set(["61-90 days", "90+ days"]);

export function deriveSignals(data: QuickBooksOverview): Signal[] {
  const out: Signal[] = [];

  // 1. Receivables that have aged past the point of "they'll get to it".
  const ar = data.ar;
  if (ar && ar.total > 0) {
    const late = ar.buckets
      .filter((b) => LATE_BUCKETS.has(b.label))
      .reduce((sum, b) => sum + b.amount, 0);
    if (late > 0) {
      const share = late / ar.total;
      const worst = [...ar.clients].sort(
        (a, b) => (b.oldest_days ?? 0) - (a.oldest_days ?? 0),
      )[0];
      out.push({
        id: "ar-late",
        severity: share >= 0.1 ? "critical" : "warn",
        headline: "Receivables have aged past 60 days",
        figure: usd(late),
        detail: worst
          ? `${Math.round(share * 100)}% of what's owed. Oldest is ${worst.client} at ${worst.oldest_days} days.`
          : `${Math.round(share * 100)}% of what's owed.`,
        goTo: "open",
      });
    }
  }

  // 2. Payables exceed the cash on hand to cover them.
  const cash = data.liquidity?.cash;
  const ap = data.ap?.total ?? 0;
  if (cash != null && cash > 0 && ap > cash) {
    out.push({
      id: "ap-over-cash",
      severity: "critical",
      headline: "You owe more than you're holding",
      figure: usd(ap - cash),
      detail: `${usd(ap)} in bills against ${usd(cash)} cash. Collections need to land before these do.`,
      goTo: "open",
    });
  }

  // 3. Clients who take far longer than everyone else to pay.
  const dso = data.dso;
  if (dso?.dso_days != null && dso.slowest_clients.length) {
    const threshold = Math.max(dso.dso_days * 1.75, 40);
    const slow = dso.slowest_clients.filter((c) => c.avg_days >= threshold);
    if (slow.length) {
      const tied = slow.reduce((sum, c) => sum + c.amount, 0);
      out.push({
        id: "slow-payers",
        severity: "warn",
        headline: `${slow.length} ${plural(slow.length, "client")} pay well after everyone else`,
        figure: usd(tied),
        detail: `${Math.round(threshold)}+ days versus a ${dso.dso_days}-day average.`,
        goTo: "clients",
      });
    }
  }

  // 4. Cost that can't be attributed to a client — the reason margins read high.
  const unattached = data.unattached_cost;
  if (unattached && unattached.unattached_pct >= 25) {
    out.push({
      id: "cost-untagged",
      severity: unattached.unattached_pct >= 50 ? "warn" : "info",
      headline: `${Math.round(unattached.unattached_pct)}% of purchases aren't tied to a client`,
      figure: usd(unattached.cost_of_service_unattached),
      detail:
        "Billable cost with nowhere to land, so per-client margin reads higher than it is.",
      goTo: "costs",
    });
  }

  // 5. Income landing outside any revenue segment.
  const coverage =
    data.revenue_by_class?.coverage_pct ?? data.class_coverage?.coverage_pct;
  const unclassified =
    data.revenue_by_class?.unclassified ?? data.class_coverage?.unclassified ?? 0;
  if (coverage != null && coverage < 90 && unclassified > 0) {
    out.push({
      id: "segment-gap",
      severity: coverage < 70 ? "warn" : "info",
      headline: "Some income isn't assigned to a segment",
      figure: usd(unclassified),
      detail: `${Math.round(coverage)}% of income is classified — the rest can't be split by line of business.`,
      goTo: "revenue",
    });
  }

  // 6. Spend concentrated in a handful of vendors.
  const ev = data.expenses_by_vendor;
  if (ev && ev.top3_concentration_pct >= 50 && ev.vendor_count > 3) {
    out.push({
      id: "vendor-concentration",
      severity: "info",
      headline: `Top 3 vendors carry ${Math.round(ev.top3_concentration_pct)}% of spend`,
      figure: usd(ev.total),
      detail: `Across ${ev.vendor_count} vendors total.`,
      goTo: "costs",
    });
  }

  // 7. Collections falling behind billing.
  const bvc = data.billing_vs_cash;
  if (bvc && bvc.invoiced_total > 0 && bvc.collection_rate_pct < 85) {
    out.push({
      id: "collection-rate",
      severity: bvc.collection_rate_pct < 70 ? "warn" : "info",
      headline: "Collections are trailing what you billed",
      figure: `${Math.round(bvc.collection_rate_pct)}%`,
      detail: `${usd(bvc.open_ar)} of this year's invoicing is still outstanding.`,
      goTo: "today",
    });
  }

  // 8. The data itself is stale or incomplete — say so rather than showing gaps.
  const failed = Object.keys(data.errors ?? {});
  if (data.sync_status === "failed" || failed.length) {
    out.push({
      id: "sync",
      severity: data.sync_status === "failed" ? "warn" : "info",
      headline:
        data.sync_status === "failed"
          ? "The last QuickBooks sync failed"
          : `${failed.length} ${plural(failed.length, "panel")} couldn't load`,
      detail: failed.length ? failed.join(", ") : "Figures below may be out of date.",
    });
  }

  return out.sort((a, b) => SEVERITY_RANK[a.severity] - SEVERITY_RANK[b.severity]);
}

/* ── self-check ─────────────────────────────────────────────────────────── */

function demo() {
  const assert = (cond: boolean, msg: string) => {
    if (!cond) throw new Error(`FAIL: ${msg}`);
  };
  const base = { errors: {}, sync_status: "ok" } as unknown as QuickBooksOverview;

  assert(deriveSignals(base).length === 0, "clean payload produces no signals");

  const late = deriveSignals({
    ...base,
    ar: {
      total: 100_000,
      invoice_count: 10,
      overdue_total: 30_000,
      buckets: [
        { label: "Not yet due", amount: 70_000 },
        { label: "90+ days", amount: 30_000 },
      ],
      clients: [{ client: "Acme", amount: 30_000, invoices: 2, oldest_days: 120 }],
    },
  } as QuickBooksOverview);
  assert(late[0].id === "ar-late", "aged AR surfaces");
  assert(late[0].severity === "critical", "30% aged is critical, not a warning");
  assert(late[0].detail!.includes("Acme"), "names the oldest debtor");

  const ordered = deriveSignals({
    ...base,
    liquidity: { as_of: "", cash: 10_000, net_cash_change: null },
    ap: { total: 50_000, bill_count: 5, buckets: [], vendors: [] },
    expenses_by_vendor: {
      total: 200_000,
      vendor_count: 20,
      top3_concentration_pct: 60,
      vendors: [],
    },
  } as unknown as QuickBooksOverview);
  assert(ordered[0].severity === "critical", "critical sorts above info");
  assert(ordered[0].id === "ap-over-cash", "payables over cash is the critical one");

  // Thresholds are exclusive at the boundary — 24% untagged is noise, 25% is not.
  const quiet = deriveSignals({
    ...base,
    unattached_cost: {
      purchase_count: 100,
      purchase_total: 0,
      unattached_count: 24,
      unattached_pct: 24,
      cost_of_service_unattached: 5_000,
      accounts: [],
    },
  } as unknown as QuickBooksOverview);
  assert(quiet.length === 0, "below-threshold untagged cost stays quiet");

  console.log("qb-signals: all checks passed");
}

if (process.argv[1]?.endsWith("qb-signals.ts")) demo();

File: backend/app/financial/qb_signals.py
"""Turns the QuickBooks overview payload into a short list of things a human
should actually do something about.

A signal only exists once a threshold is crossed, and it says so in a sentence.

Ported from frontend/src/financial/lib/qb-signals.ts. The nightly insight job and
the dashboard both read this, so the thresholds live in exactly one place.
"""

from __future__ import annotations

import math
from typing import Any

_SEVERITY_RANK = {"critical": 0, "warn": 1, "info": 2}

# QuickBooks buckets that mean genuinely late, not merely outstanding.
_LATE_BUCKETS = {"61-90 days", "90+ days"}


def js_round(value: float) -> int:
    """Round half away from zero, matching JavaScript's Math.round.

    Python's built-in round() is banker's rounding, which would render 2.5 as 2
    and silently disagree with the figures this dashboard has always shown.
    """
    return math.floor(value + 0.5) if value >= 0 else math.ceil(value - 0.5)


def usd(value: float) -> str:
    """Whole dollars with thousands separators: 14419.33 -> "$14,419"."""
    rounded = js_round(value or 0)
    return f"-${abs(rounded):,}" if rounded < 0 else f"${rounded:,}"


def _plural(count: int, one: str, many: str | None = None) -> str:
    return one if count == 1 else (many or f"{one}s")


def slow_payer_threshold(dso_days: float | None) -> float | None:
    """Days-late cutoff for a "slow payer": 1.75x the fleet average, floored at 40.

    Shared by the slow-payers signal and the chase rows so tuning one always
    tunes the other.
    """
    if dso_days is None:
        return None
    return max(float(dso_days) * 1.75, 40)


def coverage_gap(data: dict[str, Any]) -> tuple[float, float] | None:
    """Resolve (coverage_pct, unclassified) for the revenue-segment gap.

    Reads `revenue_by_class` first, falling back to `class_coverage`. Returns
    None when there's no gap worth reporting: coverage missing, coverage at or
    above 90%, or nothing unclassified.

    The `is None` checks are load-bearing — a legitimate coverage_pct of 0 must
    be used as-is, not treated as absent and fallen back from.
    """
    rbc = data.get("revenue_by_class") or {}
    cc = data.get("class_coverage") or {}
    coverage = rbc.get("coverage_pct")
    if coverage is None:
        coverage = cc.get("coverage_pct")
    unclassified = rbc.get("unclassified")
    if unclassified is None:
        unclassified = cc.get("unclassified") or 0
    if coverage is None or float(coverage) >= 90 or float(unclassified) <= 0:
        return None
    return float(coverage), float(unclassified)


def derive_signals(data: dict[str, Any]) -> list[dict[str, Any]]:
    out: list[dict[str, Any]] = []

    # 1. Receivables aged past the point of "they'll get to it".
    ar = data.get("ar") or {}
    ar_total = float(ar.get("total") or 0)
    if ar_total > 0:
        late = sum(
            float(b.get("amount") or 0)
            for b in (ar.get("buckets") or [])
            if b.get("label") in _LATE_BUCKETS
        )
        if late > 0:
            share = late / ar_total
            clients = sorted(
                ar.get("clients") or [],
                key=lambda c: c.get("oldest_days") or 0,
                reverse=True,
            )
            worst = clients[0] if clients else None
            pct = js_round(share * 100)
            detail = f"{pct}% of what's owed."
            if worst:
                detail = (
                    f"{pct}% of what's owed. "
                    f"Oldest is {worst['client']} at {worst['oldest_days']} days."
                )
            out.append({
                "id": "ar-late",
                "severity": "critical" if share >= 0.1 else "warn",
                "headline": "Receivables have aged past 60 days",
                "figure": usd(late),
                "detail": detail,
                "go_to": "open",
            })

    # 2. Payables exceed the cash on hand to cover them.
    cash = (data.get("liquidity") or {}).get("cash")
    ap_total = float((data.get("ap") or {}).get("total") or 0)
    if cash is not None and float(cash) > 0 and ap_total > float(cash):
        out.append({
            "id": "ap-over-cash",
            "severity": "critical",
            "headline": "You owe more than you're holding",
            "figure": usd(ap_total - float(cash)),
            "detail": (
                f"{usd(ap_total)} in bills against {usd(float(cash))} cash. "
                "Collections need to land before these do."
            ),
            "go_to": "open",
        })

    # 3. Clients who take far longer than everyone else to pay.
    dso = data.get("dso") or {}
    dso_days = dso.get("dso_days")
    slowest = dso.get("slowest_clients") or []
    threshold = slow_payer_threshold(dso_days)
    if threshold is not None and slowest:
        slow = [c for c in slowest if float(c.get("avg_days") or 0) >= threshold]
        if slow:
            tied = sum(float(c.get("amount") or 0) for c in slow)
            out.append({
                "id": "slow-payers",
                "severity": "warn",
                "headline": (
                    f"{len(slow)} {_plural(len(slow), 'client')} "
                    "pay well after everyone else"
                ),
                "figure": usd(tied),
                "detail": (
                    f"{js_round(threshold)}+ days versus a {dso_days}-day average."
                ),
                "go_to": "clients",
            })

    # 4. Cost that can't be attributed to a client — why margins read high.
    unattached = data.get("unattached_cost") or {}
    unattached_pct = unattached.get("unattached_pct")
    if unattached_pct is not None and float(unattached_pct) >= 25:
        out.append({
            "id": "cost-untagged",
            "severity": "warn" if float(unattached_pct) >= 50 else "info",
            "headline": (
                f"{js_round(float(unattached_pct))}% of purchases "
                "aren't tied to a client"
            ),
            "figure": usd(float(unattached.get("cost_of_service_unattached") or 0)),
            "detail": (
                "Billable cost with nowhere to land, so per-client margin "
                "reads higher than it is."
            ),
            "go_to": "costs",
        })

    # 5. Income landing outside any revenue segment.
    gap = coverage_gap(data)
    if gap is not None:
        coverage, unclassified = gap
        out.append({
            "id": "segment-gap",
            "severity": "warn" if coverage < 70 else "info",
            "headline": "Some income isn't assigned to a segment",
            "figure": usd(unclassified),
            "detail": (
                f"{js_round(coverage)}% of income is classified — "
                "the rest can't be split by line of business."
            ),
            "go_to": "revenue",
        })

    # 6. Spend concentrated in a handful of vendors.
    ev = data.get("expenses_by_vendor") or {}
    top3 = ev.get("top3_concentration_pct")
    if top3 is not None and float(top3) >= 50 and int(ev.get("vendor_count") or 0) > 3:
        out.append({
            "id": "vendor-concentration",
            "severity": "info",
            "headline": f"Top 3 vendors carry {js_round(float(top3))}% of spend",
            "figure": usd(float(ev.get("total") or 0)),
            "detail": f"Across {ev.get('vendor_count')} vendors total.",
            "go_to": "costs",
        })

    # 7. Collections falling behind billing.
    bvc = data.get("billing_vs_cash") or {}
    invoiced = float(bvc.get("invoiced_total") or 0)
    rate = bvc.get("collection_rate_pct")
    if invoiced > 0 and rate is not None and float(rate) < 85:
        out.append({
            "id": "collection-rate",
            "severity": "warn" if float(rate) < 70 else "info",
            "headline": "Collections are trailing what you billed",
            "figure": f"{js_round(float(rate))}%",
            "detail": (
                f"{usd(float(bvc.get('open_ar') or 0))} of this year's invoicing "
                "is still outstanding."
            ),
            "go_to": "today",
        })

    # 8. The data itself is stale or incomplete — say so rather than showing gaps.
    failed = list((data.get("errors") or {}).keys())
    if data.get("sync_status") == "failed" or failed:
        sync_failed = data.get("sync_status") == "failed"
        out.append({
            "id": "sync",
            "severity": "warn" if sync_failed else "info",
            "headline": (
                "The last QuickBooks sync failed"
                if sync_failed
                else f"{len(failed)} {_plural(len(failed), 'panel')} couldn't load"
            ),
            "detail": (
                ", ".join(failed) if failed else "Figures below may be out of date."
            ),
        })

    out.sort(key=lambda s: _SEVERITY_RANK[s["severity"]])
    return out
